src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Unix.Android.cs

```diff
@@ -142,6 +142,56 @@ private static TimeZoneInfo GetLocalTimeZoneCore()
             return Utc;
         }
 
+        private static TimeSpan GetCacheLocalUtcOffset(DateTime dateTime, TimeZoneInfoOptions flags)
+        {
+                CachedData cachedData = s_cachedData;
+                return cachedData.Local.GetUtcOffset(dateTime, flags, cachedData);
+        }
+
+        private static TimeSpan? _localUtcOffset;
+        private static object _localUtcOffsetLock = new();
+        private static Thread? _loadAndroidTZData;
+        // Shortcut for TimeZoneInfo.Local.GetUtcOffset
+        // On Android, loading AndroidTZData while obtaining cachedData.Local is expensive for startup.
+        // We introduce a fast result for GetLocalUtcOffset that relies on the date time offset being
+        // passed into monovm_initialize(_preparsed) from Java in seconds as LOCAL_DATE_TIME_OFFSET.
+        // However, to handle timezone changes during the app lifetime, AndroidTZData needs to be loaded.
+        // The fast path is initially used, and we start a background thread to get cachedData.Local
+        internal static TimeSpan GetLocalUtcOffset(DateTime dateTime, TimeZoneInfoOptions flags)
+        {
+            if (_localUtcOffset != null) // The background thread finished, the cache is loaded.
+                return GetCacheLocalUtcOffset(dateTime, flags);
+
+            if (_localUtcOffset == null && _loadAndroidTZData == null) // The cache isn't loaded and no background thread has been created
+            {
+                lock (_localUtcOffsetLock)
+                {
+                    // GetLocalUtcOffset may be called multiple times before a cache is loaded and a background thread is running,
+                    // once the lock is available, check for a cache and background thread.
+                    if (_localUtcOffset != null)
+                        return GetCacheLocalUtcOffset(dateTime, flags);
+
+                    if (_loadAndroidTZData == null)
+                    {
+                        _loadAndroidTZData = new Thread(() => {
+                            Thread.Sleep(1000);
```

Comment:
Like @eerhardt mentioned the 1 second here, a slow app could be more than 1 second, is there a way this can use `ManualResetEvent`? (or whatever the "good" reset event is)

Then something in xamarin/xamarin-android, maybe `Android.App.Activity.OnCreate()` could call into the runtime. @grendello do you have any thoughts how we could signal to the runtime the app is done loading?

Reply:
If the app takes more than 1s to start, then the time to load TZ doesn't really matter. There's no reason to complicate the code with signalling, especially that Android is notoriously hard in this respect... The app is done loading when the `Displayed: app` line is output to logcat, but there's no way to catch that event in the app itself. 

1s is very generous, but we can increase it to 1.5s if we think this is better. However, anything above ~700ms is perceived as "slow" by the user, so I really wouldn't worry about this.